### ISSUE TITLE
OLEditor - delete feature from WFS layer doesn't work

### ISSUE THREAD

**Issue by fegyi001** [heron-mc/heron-mc#421]
I experienced the following: 
When editing a WFS layer with OLEditor everything works fine when adding or modifying geometries. When I try to delete, however, the element disappears from the browser, but not from the database. Therefore, when I refresh the page the "deleted" feature is back again.

I found the solution for the problem. The following file has to be fixed: 'DeleteFeature.js' from ux/oleditor/ole/client/lib/Editor/Control folder. 

This should be the new content of the "deleteFeature()" method:

    deleteFeature: function () {
        if (this.layer.selectedFeatures.length > 0) {
			for(var i=0; i<this.layer.selectedFeatures.length; i++) {
				var feature = this.layer.selectedFeatures[i];
				feature.state = OpenLayers.State.DELETE;
				this.layer.events.triggerEvent('afterfeaturemodified', {feature: feature});
			}
			this.layer.destroyFeatures(this.layer.selectedFeatures);
                        this.layer.events.triggerEvent('featureunselected');
             }
    }

**Comment by justb4** [comment]
Ok, but what do you mean with "database"? The back-end database on which the WFS Layer is based, e.g. via GeoServer? Then this mean that WFS-T is involved and most probably some kind of authorization. 

It is also the first time I see that the OLEditor is used to edit WFS features! Offcourse this could be possible when you provide the OLEditor with the WFS Layer instance within the config. This sounds very advanced and possibly has many use-cases! But still for modifications (updates/deletes) to features, interaction is needed with the WFS. Keep us posted. 

The OLEditor code for Heron is maintained at https://github.com/justb4/ole. Any changes will be made there and then offered to the OLEditor master branch via Pull Requests so other users benefit as well.
 Thanks!

**Comment by fegyi001** [comment]
Yes, I mean by using WFS-T via GeoServer. I'm surprised that no one has ever tried to do it before.

Here is my code for that:

			Ext.namespace("Heron");
			Ext.namespace("Heron.options");
			Ext.namespace("Heron.options.map");
			Ext.namespace("Heron.options.layertree");
			
			OpenLayers.ProxyHost = "/cgi-bin/proxy.cgi?url=";
			Heron.globals.serviceUrl = "/cgi-bin/heron.cgi";
			
			var geoserverHost = "http://localhost:8061/geoserver/wms";

			Heron.options.map.settings = {
				projection: "EPSG:900913",
				units: "m",
				center: "2122032,6019521",
				zoom: 7,
				xy_precision: 2,
				scales: [5000000, 1000000, 500000, 100000, 50000, 10000, 5000, 1000, 500, 100],
				maxExtent: new OpenLayers.Bounds(-20191670,-10886464,20168664,14611976)
			};
			
			Heron.options.map.layers = [
                new OpenLayers.Layer.OSM()
            ];
			
			var editsLayer = new OpenLayers.Layer.Vector("Editing layer", {
                strategies: [
                    new OpenLayers.Strategy.BBOX(),
                    new OpenLayers.Strategy.Save({
                        auto: true
                    })
                ],
                projection : "EPSG:900913",
                protocol: new OpenLayers.Protocol.WFS({
                    version: "1.0.0",
                    srsName: "EPSG:900913",
                    url: geoserverHost.replace(/wms/,'wfs'),
                    featurePrefix: "heron",
                    featureType: "edits",
                    geometryName: "geom"
                }),
                visibility: true,
                styleMap: new OpenLayers.StyleMap({
                    'default': new OpenLayers.Style({
						fillColor: '#0077ff',
						fillOpacity: 0.8,
						strokeColor: '#003377',
						strokeWidth: 2,
						graphicZIndex: 1,
						pointRadius: 5
					}),
					'select': new OpenLayers.Style({
						fillColor: '#ffcc00',
						strokeColor: '#ff7700',
						graphicZIndex: 2
					}),
					'temporary': new OpenLayers.Style({
						fillColor: '#ffcc00',
						fillOpacity: 0.8,
						strokeColor: '#ff7700',
						strokeWidth: 2,
						graphicZIndex: 2,
						pointRadius: 5
					})                   
                })
            });
			
			Heron.options.map.layers.push(editsLayer);
            
            Heron.options.layertree.tree = [
                {
                    text: 'Overlays',
                    expanded: true,
                    children: [
                        {nodeType: "gx_layer", layer: "Editing layer"}
					]
                },
                {
                    text: 'Base layers',
                    expanded: true,
                    children: [
                        {nodeType: "gx_layer", layer: "OpenStreetMap"}
                    ]
                }
            ];
			
			Heron.options.map.toolbar = [
				{type: "pan"},
			    {type: "oleditor", 
			        options: {
			            pressed: true,
			            olEditorOptions: {
			                activeControls: ['Navigation', 'SnappingSettings', 'CADTools', 
								'Separator', 'DeleteFeature', 'TransformFeature', 
								'SelectFeature', 'Separator', 'ModifyFeature', 'Separator'],
			                featureTypes: ['polygon'],
			                editLayer: editsLayer,
			                styleMap: editsLayer.styleMap
			            }
			        }
			    }
			];
			
			Heron.layout = {
				xtype: 'panel',
				layout: 'border',
				border: false,
				items: [
					{
						xtype: 'hr_mappanel',
						region: 'center',
						hropts: Heron.options.map
					},
					{
						xtype: 'panel',
						title: 'Heron rocks!',
						layout: 'accordion',
						region: 'west',
						width: 270,
						collapsible: true,
						border: false,
						split: true,
						defaults: {autoScroll: true},
						layoutConfig: {
							animate: true,
							activeOnTop: false
						},
						items: [
							{
								xtype: 'hr_layertreepanel',
								region: 'west',
								collapsible: true,
								border: true,
								hropts: Heron.options.layertree
							}
						]
					}
				]
			};

**Comment by justb4** [comment]
Still my question: how do you arrange authorization, as anyone can now 
modify the database, or is this an intranet app?

**Comment by fegyi001** [comment]
The above url for GeoServer ("http://localhost:8061/geoserver/wms") is just there for an example. In real life, I use a Java proxy, and the proxy does the authentication (reads the session etc.), and adds basic authentication parameters to the GeoServer call. 

Without a proxy the basic authentication parameters should be hard-coded into the url like this: http://user:password@localhost:8061/geoserver/wms. 

The user can or cannot have 'w' rights to the given layer, this is handled on the GeoServer side.

**Comment by justb4** [comment]
Ok, smart. Thanks for sharing. The source for OLEditor is in a separate repo, a clone of OLEditor master branch: https://github.com/justb4/ole (and later probably be moved to Heron-mc organisation GH). Will integrate changes there. For future modifications you may consider using a Pull Request.

**Comment by fegyi001** [comment]
Thanks! 
How can I contribute to OLEditor in the future? I understand that I can pull the project, but which branch should I use? There are 3 now if I see correctly. Will I be able to push any modifications?

**Comment by justb4** [comment]
On 14-05-15 15:42, Gergely Padányi-Gulyás wrote:
> Thanks!
> How can I contribute to OLEditor in the future? I understand that I can
> pull the project, but which branch should I use? There are 3 now if I
> see correctly. Will I be able to push any modifications?
I think that there are 2 branches:
1. the official master branch at https://github.com/geops/ole
2. a clone for use in Heron at https://github.com/justb4/ole
(3. your clone of 2. not yet or not needed)

The master branch from 2. is used in Heron as to have a fixed working 
version and to make modifications that are immediately available to 
Heron (while e.g. waiting for a PR to 1. as now).

The/my workflow is as follows:
- for a function/bug in Heron an OLE-change is needed
- open an issue in 1. (and one in in Heron referencing this issue)
- fix the issue in 2. with commit msg referencing the issue #
- create a PR in 2 to be offered to 1

 From time to time I merge back any changes in 1. into 2 like just now 
for new ContextMenu feature.

You have 3 options:
- You could work in the same way either using your own clone (3)  and do 
the same workflow, only now your upstream master is 2 so after fixing 
you issue a PR from 3. to  and I do a PR from 2 to 1.
- You do the same as me and issue PRs directly to 1 and ask me to merge back
- I can make you a committer for 2. so you can work more directly as in 
the workflow above.

If you plan to make regular changes, possibly the 3rd option would be 
the easiest. I even plan to move 2. to the Heron MC organization in 
GitHub https://github.com/heron-mc
